src/ContainerTrait.php

```diff
@@ -14,22 +39,152 @@ public function _unique_element($desired = null)
         return $desired.($postfix > 1 ? ('_'.$postfix) : '');
     }
 
-    protected function _unique() {
+    /**
+     * Extension to add() method which will perform linking of
+     * the object with the current class.
+     */
+    protected function _add_Container($element, $args = [])
+    {
+        // Normalize the arguments, bring name out
+        if (is_string($args)) {
+
+            // passed as string
+            $args=[$args];
+        } elseif (!is_array($args)) {
+
+            throw new Exception(['Second argument must be array','arg2'=>$args]);
+        } elseif (is_array($args) && isset($args['name'])) {
+
+            // passed as ['name'=>'foo'];
+            $args[0]=$args['name'];
+            unset($args['name']);
+        } elseif (isset($element->short_name)) {
+
+            // element has a name already
+            $args[0]=$element->short_name;
+        } elseif (isset($element->_trackableTrait)) {
+
+            // ask element on his preferred name, then make it unique.
+            $cn = $element->getDesiredName();
+            $args[0] = $this->_unique_element($cn);
+        } else {
+
+            // generate name based on the class
+            $cn = str_replace('\\', '_', strtolower(get_class($element)));
+            $args[0] = $this->_unique_element($cn);
+        }
+
+        // Maybe element already exists
+        if (isset($this->elements[$args[0]])) {
+            throw new Exception([
+                'Element with requested name already exists',
+                'element'=>$element,
+                'name'=>$args[0],
+                'this'=>$this,
+                'arg2'=>$args
+            ]);
+        }
+
+        if(isset($this->_appScope) && isset($element->_appScope)) {
+            $element->app = $this->app;
+        }
+
+        $element->owner = $this;
+        $element->short_name = $args[0];
```

Comment:
Yes, i thought about that too. but sometimes we want to add object which will not be tracked.

In this cases - what do we actually do? there is no point calling this method at all.  So I returned those lines. 

Maybe if element is not trackable and it is passed to _add_Container we should just spit it out right away?

The other option is that the element still wants some unique identifier, which is what we actually do here and ATK would want it like this. The alternative is to add UnTrackableTrait. Let's leave until we have more suggestions here.